File: src/test/java/com/runninglane/dto/buddy/test/java/cases/contract/GetterSetterBehaviorContractTest.java
package com.runninglane.dto.buddy.test.java.cases.contract;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.List;

import static com.runninglane.dto.buddy.test.java.cases.contract.CompliantBaseClasses.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

/**
 * The rule is the returned class must be a concrete class
 */
public class GetterSetterBehaviorContractTest {
    private final DtoBuddy dtoBuddy = new DtoBuddy();

    private void test(Class<?> baseClass, boolean expectBaseClassReturned, boolean testGetSet, boolean testGetSetInJavaStyle) {
        try {
            Class<?> concreteClass = dtoBuddy.implement(baseClass);
            assertFalse(concreteClass.isInterface());
            assertFalse(Modifier.isAbstract(concreteClass.getModifiers()));
            if (expectBaseClassReturned) {
                assertEquals(baseClass, concreteClass);
            }
            if (testGetSet) {
                TestGetSetCapability.testGetSet(dtoBuddy, concreteClass);
            }
        } catch (Throwable e) {
            throw new RuntimeException("Failed to test " + baseClass.getSimpleName(), e);
        }
    }

    @Test
    public void testInterface() {
        List<Class<?>> classes = Arrays.asList(
            InterfaceWithNoMember.class,
            InterfaceWithAbstractGetter.class,
            InterfaceWithAbstractGetterAndAbstractSetter.class,
            InterfaceWithAbstractProperty.class
        );

        for (Class<?> baseClass : classes) {
            boolean testGetSet = baseClass != InterfaceWithNoMember.class;
            boolean testGetSetInJavaStyle = testGetSet && !baseClass.getSimpleName().contains("Property");
            test(baseClass, false, testGetSet, testGetSetInJavaStyle);
        }
    }

    @Test
    public void testAbstractClass() {
        List<Class<?>> classes = Arrays.asList(
            AbstractClassWithNoMember.class,
            AbstractClassWithAbstractGetter.class,
            AbstractClassWithAbstractGetterAndAbstractSetter.class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithoutField.class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithField.class,
            AbstractClassWithAbstractImmutableProperty.class,
            AbstractClassWithAbstractMutableProperty.class,
            AbstractClassWithConcreteMutableProperty.class
        );

        for (Class<?> baseClass : classes) {
            boolean testGetSet = !baseClass.getSimpleName().matches(".*(NoMember|Concrete).*");
            boolean testGetSetInJavaStyle = testGetSet && !baseClass.getSimpleName().contains("Property");
            test(baseClass, false, testGetSet, testGetSetInJavaStyle);
        }
    }

    @Test
    public void testConcreteClass() {
        List<Class<?>> classes = Arrays.asList(
            ConcreteClassWithNoMember.class,
            ConcreteClassWithGetterAndSetterWithoutField.class,
            ConcreteClassWithGetterAndSetterWithField.class
        );

        for (Class<?> baseClass : classes) {
            test(baseClass, true, false, false);
        }
    }

}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/contract/GetterSetterBehaviorContractTest.kt
package com.runninglane.dto.buddy.test.cases.contract

import com.runninglane.dto.buddy.DtoBuddy
import kotlin.reflect.KClass
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse

/**
 * The rule is the returned class must be a concrete class
 */
class GetterSetterBehaviorContractTest : TestGetSetCapability {
    private val dtoBuddy = DtoBuddy()

    private fun test(baseClass: KClass<*>, expectBaseClassReturned: Boolean, testGetSet: Boolean, testGetSetInJavaStyle: Boolean) {
        try {
            val concreteClass = dtoBuddy.implement(baseClass)
            assertFalse(concreteClass.java.isInterface)
            assertFalse(concreteClass.isAbstract)
            if (expectBaseClassReturned) {
                assertEquals(baseClass, concreteClass)
            }
            if (testGetSet) {
                testGetSet(dtoBuddy, concreteClass, testGetSetInJavaStyle)
            }
        } catch (e: Throwable) {
            throw RuntimeException("Failed to test ${baseClass.simpleName}", e)
        }
    }

    @Test
    fun testInterface() {
        listOf(
            InterfaceWithNoMember::class ,
            InterfaceWithAbstractGetter::class ,
            InterfaceWithAbstractGetterAndAbstractSetter::class ,
            InterfaceWithAbstractProperty::class ,
        ).forEach { baseClass ->
            val testGetSet = baseClass != InterfaceWithNoMember::class
            val testGetSetInJavaStyle = testGetSet && !baseClass.simpleName!!.contains("Property")
            test(baseClass, false, testGetSet, testGetSetInJavaStyle)
        }
    }

    @Test
    fun testAbstractClass() {
        listOf(
            AbstractClassWithNoMember::class,
            AbstractClassWithAbstractGetter::class,
            AbstractClassWithAbstractGetterAndAbstractSetter::class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithoutField::class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithField::class,
            AbstractClassWithAbstractImmutableProperty::class,
            AbstractClassWithAbstractMutableProperty::class,
            AbstractClassWithConcreteMutableProperty::class
        ).forEach { baseClass ->
            val testGetSet = baseClass.simpleName?.matches(Regex(".*(NoMember|Concrete).*")) != true
            val testGetSetInJavaStyle = testGetSet && !baseClass.simpleName!!.contains("Property")
            test(baseClass, expectBaseClassReturned = false, testGetSet, testGetSetInJavaStyle)
        }
    }

    @Test
    fun testConcreteClass() {
        listOf(
            ConcreteClassWithNoMember::class,
            ConcreteClassWithGetterAndSetterWithoutField::class,
            ConcreteClassWithGetterAndSetterWithField::class,
            ConcreteClassWithConcreteMutableProperty::class
        ).forEach { baseClass ->
            test(baseClass, expectBaseClassReturned = true, testGetSet = false, testGetSetInJavaStyle = false)
        }
    }

}